common/objectcachev2.h

```diff
@@ -36,154 +37,138 @@ class ObjectCacheV2 {
+    intrusive_list<Box> lru_list;
     uint64_t lifespan;
     photon::Timer _timer;
     bool _exit = false;
 
-    std::shared_ptr<V> __update(Box& item, std::shared_ptr<V> val) {
-        auto ret_val = val;
-        auto old_val = item.writer().update(val, photon::now);
-        reclaim_queue.template send<PhotonPause>(old_val);
-        return ret_val;
-    }
-
-    std::shared_ptr<V> __update(Box& item, V* val) {
-        std::shared_ptr<V> ptr(val);
-        return __update(item, ptr);
-    }
-
     template <typename KeyType>
-    Box& __find_or_create_item(KeyType&& key) {
-        Box keyitem(key);
-        Box* item = nullptr;
+    Box& __find_or_create_box(KeyType&& key) {
+        Box keybox(key);
+        Box* box = nullptr;
         SCOPED_LOCK(maplock);
-        auto it = map.find(keyitem);
+        auto it = map.find(keybox);
         if (it == map.end()) {
-            // item = new Box(std::forward<KeyType>(key));
-            auto rt = map.emplace(key);
-            if (rt.second) item = (Box*)&*rt.first;
+            auto rt = map.emplace(std::forward<KeyType>(key));
+            if (rt.second) box = (Box*)&*rt.first;
         } else
-            item = (Box*)&*it;
-        assert(item);
-        item->timestamp = photon::now;
-        cycle_list.pop(item);
-        cycle_list.push_back(item);
-        return *item;
+            box = (Box*)&*it;
+        assert(box);
+        lru_list.pop(box);
+        box->acquire();
+        return *box;
     }
-
     uint64_t __expire() {
-        intrusive_list<Box> delete_list;
+        std::vector<std::shared_ptr<V>> to_release;
         uint64_t now = photon::now;
+        uint64_t reclaim_before = photon::sat_sub(now, lifespan);
         {
             SCOPED_LOCK(maplock);
-            if (cycle_list.empty()) return 0;
-            if (photon::sat_add(cycle_list.front()->timestamp, lifespan) >
-                now) {
-                return photon::sat_add(cycle_list.front()->timestamp,
-                                       lifespan) -
-                       now;
+            if (lru_list.empty()) return 0;
+            if (lru_list.front()->timestamp > reclaim_before) {
+                return lru_list.front()->timestamp - reclaim_before;
             }
-            auto x = cycle_list.front();
-            while (x && (photon::sat_add(x->timestamp, lifespan) < now)) {
-                cycle_list.pop(x);
-                __update(*x, nullptr);
-                map.erase(*x);
-                x = cycle_list.front();
+            auto x = lru_list.front();
+            // here requires lock dance for lru_list and kv
+            while (x && x->timestamp < reclaim_before) {
+                lru_list.pop(x);
+                if (x->rc == 0) {
+                    // make vector holds those shared_ptr
+                    // prevent object destroy in critical zone
+                    to_release.push_back(x->clear());
+                    map.erase(*x);
```

Comment:
use map.extract().value() obtain the Box element, so as to use the intrusive_list for to_release.

Reply:
Seems add macro to check c++ version to show different implementation for those above C++17 is a good idea.

Reply:
Seem that C++17 node_handle owns the element Box, can not get out the pointer to Box without hacking.